Mathlib/RingTheory/Localization/AtPrime.lean

```diff
@@ -0,0 +1,274 @@
+/-!
+# Localizations of commutative rings at the complement of a prime ideal
+
+## Main definitions
+
+ * `is_localization.at_prime (I : ideal R) [is_prime I] (S : Type*)` expresses that `S` is a
+   localization at (the complement of) a prime ideal `I`, as an abbreviation of
+   `is_localization I.prime_compl S`
+
+## Main results
+
+ * `is_localization.at_prime.local_ring`: a theorem (not an instance) stating a localization at the
+   complement of a prime ideal is a local ring
+
+## Implementation notes
+
+See `src/ring_theory/localization/basic.lean` for a design overview.
+
+## Tags
+localization, ring localization, commutative ring localization, characteristic predicate,
+commutative ring, field of fractions
+-/
+
+
+variable {R : Type _} [CommSemiring R] (M : Submonoid R) (S : Type _) [CommSemiring S]
+
+variable [Algebra R S] {P : Type _} [CommSemiring P]
+
+section AtPrime
+
+variable (I : Ideal R) [hp : I.IsPrime]
+
+namespace Ideal
+
+/-- The complement of a prime ideal `I ⊆ R` is a submonoid of `R`. -/
+def primeCompl : Submonoid R where
+  carrier := (Iᶜ : Set R)
+  one_mem' := by convert I.ne_top_iff_one.1 hp.1
+  mul_mem' {x y} hnx hny hxy := Or.casesOn (hp.mem_or_mem hxy) hnx hny
+#align ideal.prime_compl Ideal.primeCompl
+
+theorem primeCompl_le_nonZeroDivisors [NoZeroDivisors R] : I.primeCompl ≤ nonZeroDivisors R :=
+  le_nonZeroDivisors_of_noZeroDivisors <| not_not_intro I.zero_mem
+#align ideal.prime_compl_le_non_zero_divisors Ideal.primeCompl_le_nonZeroDivisors
+
+end Ideal
+
+/-- Given a prime ideal `P`, the typeclass `is_localization.at_prime S P` states that `S` is
+isomorphic to the localization of `R` at the complement of `P`. -/
+protected abbrev IsLocalization.AtPrime :=
+  IsLocalization I.primeCompl S
+#align is_localization.at_prime IsLocalization.AtPrime
+
+/-- Given a prime ideal `P`, `localization.at_prime S P` is a localization of
+`R` at the complement of `P`, as a quotient type. -/
+protected abbrev Localization.AtPrime :=
+  Localization I.primeCompl
+#align localization.at_prime Localization.AtPrime
+
+namespace IsLocalization
+
+theorem AtPrime.Nontrivial [IsLocalization.AtPrime S I] : Nontrivial S :=
+  nontrivial_of_ne (0 : S) 1 fun hze => by
+    rw [← (algebraMap R S).map_one, ← (algebraMap R S).map_zero] at hze
+    obtain ⟨t, ht⟩ := (eq_iff_exists I.primeCompl S).1 hze
+    have htz : (t : R) = 0 := by simpa using ht.symm
+    exact t.2 (htz.symm ▸ I.zero_mem : ↑t ∈ I)
+#align is_localization.at_prime.nontrivial IsLocalization.AtPrime.Nontrivial
+
+-- Porting Note: Lean cannot find synthesization order for instance, so I turned it off
+-- not sure if this is a good idea though
+set_option synthInstance.checkSynthOrder false in
+attribute [local instance] AtPrime.Nontrivial
+
+theorem AtPrime.localRing [IsLocalization.AtPrime S I] : LocalRing S :=
+  -- Porting Note : since I couldn't get local instance running, I just specify it manually
+  letI := AtPrime.Nontrivial S I
+  LocalRing.of_nonunits_add
```

Comment:
it complains `cannot find synthesization order for instance`, I tried to mark various parameter as `semiOutParam` or `OutParam`, but didn't work